Add gui support to getcellpixels

mikoto2000 opened the pull request:
#16004 and #16047 only supported Unix terminal, but this time we added support for gVim.

mikoto2000 committed:
Added windows function.

mikoto2000 committed:
Add support on gui.

mikoto2000 committed:
Move cellsize struct.

mikoto2000 committed:
Fixed mch_calc_cell_size in os_win32.c.

mikoto2000 committed:
Add test.

zeertzjq commented on src/testdir/test_functions.vim:
```diff
@@ -4180,10 +4180,29 @@ func Test_getcellpixels()
+
 " Test for getcellpixels() on gVim
 func Test_getcellpixels_gui()
   " Not yet Windows-compatible
```
This comment should be removed

mikoto2000 replied:
Fixed in e23f96a3920c3314bbbc2bb70afd3d0a3f149539

zeertzjq commented on src/testdir/test_functions.vim:
```diff
@@ -4180,10 +4180,29 @@ func Test_getcellpixels()
+" Test for getcellpixels() for windows system
+" Windows terminal vim is not support. check return `[]`.
+func Test_getcellpixels_for_windows()
+  " Not yet Windows-compatible
```
This comment should also be removed

mikoto2000 replied:
Fixed in e23f96a3920c3314bbbc2bb70afd3d0a3f149539

zeertzjq commented on src/testdir/test_functions.vim:
```diff
@@ -4160,9 +4160,9 @@ func Test_slice()
 " Pixel size of a cell is terminal-dependent, so in the test, only the list and size 2 are checked.
-func Test_getcellpixels()
+func Test_getcellpixels_for_unix()
   " Not yet Windows-compatible
```
and maybe this as well

mikoto2000 replied:
Fixed in e23f96a3920c3314bbbc2bb70afd3d0a3f149539

ychin commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,41 @@ f_get(typval_T *argvars, typval_T *rettv)
+    {
+#endif
+        struct cellsize cs;
+        mch_calc_cell_size(&cs);
```
I think it might be easier if you just do something like this, and don't define this function in `os_win32.c`:

```c
#if defined(UNIX)
    mch_calc_cell_size(&cs);
#else
    cs_out->cs_xpixel = -1;
    cs_out->cs_ypixel = -1;
#endif
```

The reason is that outside of Unix and Win32 we have platform definitions for things like VMS / Amiga. They aren't used that often or checked in CI, but unless you define this function for all platforms, this code won't build on all of them. E.g. if you look at `mch_get_shellsize` for example, it's defined for Amiga, VMS, etc. I'm not sure if you want to define all of them just to have them set xpixel to -1.

mikoto2000 replied:
Fixed in 44b5e6cdbc53c85377a7f8e4b90b8767551df217

mikoto2000 committed:
Fix expected value in getcellpixels on gVim.

mikoto2000 committed:
Delete the incorrect comment.

mikoto2000 committed:
Remove mch_calc_cell_size in os_win32.c.

mikoto2000 committed:
floating point exception in GVim

Problem:  floating point exception in GVim
Solution: Return early when returned pixel size is 0

Signed-off-by: Christian Brabandt <cb@256bit.org>

k-takata commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,47 @@ f_get(typval_T *argvars, typval_T *rettv)
+        list_append_number(rettv->vval.v_list, (varnumber_T)cs.cs_ypixel);
+#if defined(FEAT_GUI)
+    }
+#endif
```
Your code is currently:
```C
#if defined(FEAT_GUI)
    if (...)
    {
        ...
    }
    else
    {
#endif
        ...
#if defined(FEAT_GUI)
    }
#endif
```
But I think that the following style is better:
```C
#if defined(FEAT_GUI)
    if (...)
    {
        ...
    }
    else
#endif
    {
        ...
    }
```
This can reduce the ifdefs.

mikoto2000 replied:
Fixed in 7818a497285a0c416703698c4c20cc3006acbe75.

mikoto2000 committed:
Updated the documentation for `getcellpixels()`.

Added information about Retina scaling.

k-takata commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,47 @@ f_get(typval_T *argvars, typval_T *rettv)
+    {
+#endif
+        struct cellsize cs;
+# if defined(UNIX)
```
```suggestion
#if defined(UNIX)
```
Spaces are required only when #ifs are nested.

mikoto2000 replied:
Fixed in 7818a497285a0c416703698c4c20cc3006acbe75.

k-takata commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,47 @@ f_get(typval_T *argvars, typval_T *rettv)
+        struct cellsize cs;
+# if defined(UNIX)
+        mch_calc_cell_size(&cs);
+# else
```
```suggestion
#else
```

mikoto2000 replied:
Fixed in 7818a497285a0c416703698c4c20cc3006acbe75.

k-takata commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,47 @@ f_get(typval_T *argvars, typval_T *rettv)
+        // Non-Unix CUIs are not supported, so set this to -1x-1.
+        cs.cs_xpixel = -1;
+        cs.cs_ypixel = -1;
+# endif
```
```suggestion
#endif
```

mikoto2000 replied:
Fixed in 7818a497285a0c416703698c4c20cc3006acbe75.

mikoto2000 committed:
Fixed ifdefs style.

k-takata commented on src/evalfunc.c:
```diff
@@ -5216,6 +5211,45 @@ f_get(typval_T *argvars, typval_T *rettv)
+#if defined(FEAT_GUI)
+    if (gui.in_use)
+    {
+        // success pixel size and no gui.
```
The comment "no gui" looks wrong.
